test/TESTING_WITH_DOCKER.md

```diff
@@ -0,0 +1,131 @@
+
+
+## Clone the project
+
+Use git to clone the [newrelic/newrelic-ruby-agent](https://github.com/newrelic/newrelic-ruby-agent)
+project.
+
+The [Dockerfile](../Dockerfile) and [docker-compose.yml](../docker-compose.yml)
+files are located in the root of the project, and this `TESTING_WITH_DOCKER.md`
+document is located within the `test` subdirectory of the project.
+
+
+## Using just the Dockerfile (unit tests only)
+
+The project `Dockerfile` can be used by itself to run the project unit tests.
+Docker Compose and the project `docker-compose.yml` file will be needed to
+run functional tests which involve communicating with data systems such as
+PostgreSQL and Redis.
+
+To only run the unit tests using `Dockerfile` by itself, first change
+directories to the root of the project, then build an image, and
+finally run a container from the image:
+
+```shell
+$ cd /path/to/project/git/clone
+$ docker build -t newrelic/newrelic-ruby-agent-tester .
+$ docker run -it newrelic/newrelic-ruby-agent-tester
+```
+
+* `build -t <TAG>` applies a tag to the image during building
+* `.` indicates "here" and tells Docker that the `Dockerfile` file can be found
+  in the current directory
+* `run -it` tells Docker to run interactively (`i`) and with a pseudo TTY (`t`)
+
+
+## Using Docker Compose (required for functional tests)
+
+Docker Compose launches multiple containers simultaneously to support the
+running of the functional tests that require a variety of data handling
+server applications such as PosgreSQL, Redis, memcached, etc. Each one of
+these server applications uses its own container and then there's a Ruby
+container (referred to as the "app" container) that runs the Minitest tests
+while connecting to the other containers.
+
+In one shell session, change to the root of the project and then bring up all
+containers with Docker Compose:
+
+```shell
+$ docker-compose up
+```
+
+In a separate shell session (probably in a separate terminal split, tab, or
+window), execute the "multiverse" functional tests:
+
+```shell
+$ docker-compose exec app bundle exec rake test:multiverse
+```
+
+For multiverse related arguments or to specify only a subset of tests to be ran,
+see the [test/multiverse/README.md](multiverse/README.md) doc for instructions.
+
+
+## Debugging
+
+When using Docker Compose, one shell session will produce STDOUT output that
+pertains to all of the services (MySQL, MongoDB, etc.) and the other shell
+session will produce STDOUT output related to the Ruby based functional tests.
+Both streams of output may provide information about any errors or warnings
+that take place.
+
+In lieu of running the functional tests, an interactive Bash shell can be
+launched against the running Ruby app container for Ruby and/or Linux
+debugging. While the `docker-compose up` shell session is still running,
+bring up an additional shell session and run the following:
+
+```shell
+$ docker-compose exec app bash
+```
+
```

Comment:
Are we able to connect the agent running on Docker to a playground application?

Reply:
We ought to be able to pull that off. I think there's a possibility that referring to the host is done differently between Docker for macOS and for Linux, but I'd love to give it a try.

Reply:
Created #926 for this.